Fix binary size of OpsAlreadyMovedToC10.cpp

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26237 Fix binary size of OpsAlreadyMovedToC10.cpp**

Calling a lot of `std::string` constructors is horrible for binary size, see t53997334.

Using `const char*` instead should make the binary size much smaller.

Differential Revision: [D17386002](https://our.internmc.facebook.com/intern/diff/D17386002/)

smessmer committed:
Fix binary size of OpsAlreadyMovedToC10.cpp

Calling a lot of `std::string` constructors is horrible for binary size, see t53997334.

Using `const char*` instead should make the binary size much smaller.

Differential Revision: [D17386002](https://our.internmc.facebook.com/intern/diff/D17386002/)

smessmer committed:
Update on "Fix binary size of OpsAlreadyMovedToC10.cpp"


Calling a lot of `std::string` constructors is horrible for binary size, see t53997334.

Using `const char*` instead should make the binary size much smaller.

Differential Revision: [D17386002](https://our.internmc.facebook.com/intern/diff/D17386002/)

dzhulgakov reviewed (approved):
sure, but I'd guess we shouldn't even compile it on mobile in the first place since there's only one dispatcher?

smessmer committed:
Update on "Fix binary size of OpsAlreadyMovedToC10.cpp"


Calling a lot of `std::string` constructors is horrible for binary size, see t53997334.

Using `const char*` instead should make the binary size much smaller.

Differential Revision: [D17386002](https://our.internmc.facebook.com/intern/diff/D17386002/)

smessmer committed:
Update on "Fix binary size of OpsAlreadyMovedToC10.cpp"


Calling a lot of `std::string` constructors is horrible for binary size, see t53997334.

Using `const char*` instead should make the binary size much smaller.

Differential Revision: [D17386002](https://our.internmc.facebook.com/intern/diff/D17386002/)

apaszke commented on aten/src/ATen/templates/OpsAlreadyMovedToC10.cpp:
```diff
@@ -1,25 +1,46 @@
+struct OpNameHash final {
+  size_t operator()(const std::pair<const char*, const char*>& p) const {
+      // use std::hash<std::string> because std::hash<const char*> would hash pointers and not pointed-to strings
+      return std::hash<std::string>()(p.first) ^ (~ std::hash<std::string>()(p.second));
```
I understand that we can't use the `const char*` hasher here, but doesn't this mean that each attempt to hash will actually have to allocate `std::string`s for both values, and copy their contents, only to throw them away once they're hashed? This seems quite expensive. We already have `hash_combine`, so maybe it's better to use this instead?

smessmer committed:
Update on "Fix binary size of OpsAlreadyMovedToC10.cpp"


Calling a lot of `std::string` constructors is horrible for binary size, see t53997334.

Using `const char*` instead should make the binary size much smaller.

Differential Revision: [D17386002](https://our.internmc.facebook.com/intern/diff/D17386002/)
